RFC HAL I2C: Reference implementations for K64F and STM32F0

scartmell-arm opened the pull request:
### Description

This PR contains reference implementations of the I2C HAL RFC changes for the FRDM-K64F and NUCLEO_F070RB boards.

### Tasks

- [X] K64F Sync
- [X] STM32 Sync
- [X] STM32 Async
- [X] Driver API Sync
- [ ] Driver API Async

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [X] Breaking change

Commit:
docs(hal-i2c): Add initial design document for the I2C HAL API overhaul

Commit:
docs(hal-i2c): Add modifications for timeout and multimaster support

Commit:
docs(hal-i2c): Update the RFC to reflect removing the DMAUsage parameter

Commit:
docs(hal-i2c): Add missing stop and last parameters to i2c_read/write

Commit:
docs(hal-i2c): Move address parameter into a seperate i2c_slave function

Commit:
docs(hal-i2c): Add additional defined function behaviours

Commit:
feat(hal-i2c): Add asynchronous HAL I2C implementation for STM32

Commit:
docs(hal-i2c): Make additional amendments to I2C API specification

- Add return value to i2c_frequency indicating configured frequency.
- Change return type of i2c_write/i2c_read from int to int32_t
- Add type suffix to i2c_slave_status enum.
- Rename i2c_slave_receive to i2c_slave_status.

Commit:
docs(hal-i2c): Add additional notes to synchronous read/write

Commit:
docs(hal-i2c): Add clock stretching field to target capabilities

Commit:
feat(hal-i2c): Update the HAL I2C header file to new API

Commit:
feat(hal-i2c): Disable I2C on all targets in preparation

Commit:
feat(hal-i2c): Add synchronous HAL I2C implementation for K64F

Commit:
feat(hal-i2c): Update I2C driver API to new specification

Commit:
feat(hal-i2c): Add synchronous HAL I2C implementation for STM32

Commit:
docs(hal-i2c): Add context pointer to async transfer function

Commit:
docs(hal-i2c): Fix return type of i2c_frequency function

Commit:
feat(hal-i2c): Remove unused DMAUsage member from I2C class

cmonr commented:
@scartmell-arm Something about this PR's comit history and files changed does not look right. Could you double check?

Commit:
fix(hal-i2c): Fix issue with master byte read/write failing on K64F

Remove byte_read/byte_write functions and use block read and writes for
single byte accesses.

Commit:
fix(hal-i2c): Add missing i2c_free function to API

Commit:
docs(hal-i2c): Move design document to the updated folder

cmonr commented:
@mbed-os-maintainers @ARMmbed/mbed-os-hal Readt for review when able.

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -791,7 +905,12 @@ int i2c_read(i2c_t *obj, int address, char *data, int length, int stop)
+#else
+            const bool is_slave = false;
+#endif
+            i2c_init(obj, obj_s->sda, obj_s->scl, is_slave);
```
since we have timeout here shouldn't return `I2C_ERROR_NO_SLAVE` here?

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -845,7 +971,12 @@ int i2c_write(i2c_t *obj, int address, const char *data, int length, int stop)
+#else
+            const bool is_slave = false;
+#endif
+            i2c_init(obj, obj_s->sda, obj_s->scl, is_slave);
```
since we have timeout here shouldn't return `I2C_ERROR_NO_SLAVE` here?

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -568,24 +674,25 @@ int i2c_byte_write(i2c_t *obj, int data)
 }
 
-int i2c_stop(i2c_t *obj)
+bool i2c_stop(i2c_t *obj)
```
function return type mismatch, returns I2C_ERROR_BUS_BUSY instead true/false

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -258,7 +258,77 @@ void i2c_sw_reset(i2c_t *obj)
+        return;
+    }
+
+    capabilities->minimum_frequency = 1;
```
`minimum_frequency` doesn't match what is stated in `i2c_frequency` function
min frequency is 100000 not 1
```c++
    ...
    MBED_ASSERT((frequency == 100000) || (frequency == 400000) || (frequency == 1000000));
   ..
```

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -361,16 +448,17 @@ void i2c_frequency(i2c_t *obj, int hz)
     /*  Only predefined timing for below frequencies are supported */
-    MBED_ASSERT((hz == 100000) || (hz == 400000) || (hz == 1000000));
-    handle->Init.Timing = get_i2c_timing(hz);
+    MBED_ASSERT((frequency == 100000) || (frequency == 400000) ||
```
Shouldn't it be whole range allowed here `minimum_frequency <= frequency <= maximum_frequency` ?,
And the `frequency` is then mapped to nearest feasible or error is returned.
The API description for this function suggests that the frequency set may differ of what selected
```
 *  @returns The actual frequency that the peripheral will be generating to
 *           allow a user adjust its strategy in case the target cannot be
 *           reached.
```

maciejbocianski commented on drivers/I2CSlave.cpp:
```diff
@@ -34,39 +33,43 @@ void I2CSlave::frequency(int hz)
 {
     int addr = (address & 0xFF) | 1;
-    i2c_slave_address(&_i2c, 0, addr, 0);
+
```
The line: 
`int addr = (address & 0xFF) | 1;`
breaks 10bit addressing

maciejbocianski reviewed (commented):
I have additional question to all HAL and DRIVERS API functions.
Whenever we passes device address it should be shifted left by 1 bit and filled by 1 or 0 (or LSB bit is ignored) depending of situation, it's valid for both 7bit and 10bit addressing?

Commit:
fix(hal-i2c): Fix incorrect minimum frequency in STM I2C implementation

Commit:
fix(hal-i2c): Fix STM implementation calling event handler too early

An async transfer involving a write and a read is issue in two stages,
the write is invoked, then in the write complete callback the read is
invoked followed by the read complete callback.

Currently the read and write complete callbacks invoke the HAL API event
handler, but this will be incorrectly called halfway through the
transfer after the write is completed.

To fix, only call the HAL event handler in the write complete callback
if there is no pending read.

Commit:
fix(hal-i2c): Return the correct number of transferred bytes in STM32

The STM32 attribute XferCount does not contain the number of transferred
bytes, it contains the number of bytes remaining to be sent. Calculate
the number of bytes that have been sent by subtracting this number from
the length of the respective buffer.

To fix, add an additional attribute to the asynchronous event object.
For transfers that include sending and receiving data only counting the
number of bytes sent isn't sufficient as unlike SPI the size of the send
and receive buffer may be different.

Commit:
refactor(hal-i2c): Rename HAL slave status enum to match Driver API

Commit:
fix(hal-i2c): Fix incorrect byte counts during aborted R/W transfers

The async event callback returns the number of bytes sent and received
during a transfer. For failed transfers that involve a write and read we
don't easily know which part of the transfer fails, only the number of
bytes remaining in the current operation.

To fix, set a flag when the write completes successfully, a failure
when this flag is set means the read stopped early and we can calculate
the number of bytes sent for each operation.

Commit:
refactor(hal-i2c): Change STOP function to return success or failure

Commit:
fix(hal-i2c): Add missing semicolon to STM32 I2C attribute 'tx_count'

Commit:
fix(hal-i2c): STM: Return the num of bytes sent for incomplete transfers

Commit:
refactor(hal-i2c): Remove return status from I2C start/stop functions

The return status doesn't make sense in this context, and most platforms
can't detect failure at this point. Platforms that do support detecting
failure here can handle it during read/write anyway.

cmonr commented:
@scartmell-arm You have questions/comments waiting for your attention.

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -361,16 +460,29 @@ void i2c_frequency(i2c_t *obj, int hz)
 
     // Enable the Fast Mode Plus capability
-    if (hz == 1000000) {
+    if (frequency == 1000000) {
```
shouldn't be `selected_frequency` used here?

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -431,7 +543,27 @@ void i2c_frequency(i2c_t *obj, int hz)
+}
+
+void i2c_timeout(i2c_t *obj, uint32_t timeout)
+{
```
seems the timeout parameter is not used nor saved.
Shouldn't we save it for later use in the relevant read / write functions ?

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -431,7 +543,27 @@ void i2c_frequency(i2c_t *obj, int hz)
+    handle->Init.AddressingMode  = I2C_ADDRESSINGMODE_7BIT;
+    handle->Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    handle->Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+    handle->Init.NoStretchMode   = (timeout == 0) ? I2C_NOSTRETCH_DISABLE
```
I haven't seen in header file what (timeout == 0 ) should mean ... does it mean that application wants to disable stretch or that application does not want any timeout ? 

If application wants to disable stretch support, it should use "I2C_NOSTRETCH_ENABLE" which will enable (set to 1) the bit in register to select NOSTRETCH configuraiton. I feel like this is coded the other way around right now.
I2C_NOSTRETCH_ENABLE => means NO STRETCH
I2C_NOSTRETCH_DISABLE => means STRETCH is OK  (NOT NOT ....)

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -568,46 +693,45 @@ int i2c_byte_write(i2c_t *obj, int data)
         /*  re-init slave when stop is requested */
-        i2c_init(obj, obj_s->sda, obj_s->scl);
-        return 0;
+        i2c_init(obj, obj_s->sda, obj_s->scl, is_slave);
```
In case we're slave we need to return here I think
e.g. need to replace return 0; by return;

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -831,36 +967,45 @@ int i2c_write(i2c_t *obj, int address, const char *data, int length, int stop)
-        }
-    } else {
-        DEBUG_PRINTF("ERROR in i2c_read\r\n");
+    uint32_t timeout = (BYTE_TIMEOUT_US * (length + 1));
```
This is the default timeout, but shouldn't it be the timeout defined by application using i2c_timeout ?

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -877,6 +1022,23 @@ void HAL_I2C_MasterTxCpltCallback(I2C_HandleTypeDef *hi2c)
         obj_s->event = I2C_EVENT_TRANSFER_COMPLETE;
+
+#if DEVICE_I2C_ASYNCH
+        /* */
```
Why an empty comment ?

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -885,9 +1047,27 @@ void HAL_I2C_MasterRxCpltCallback(I2C_HandleTypeDef *hi2c)
+
+    obj->handler(obj, &event, obj->ctx);
+
+    obj->handler = NULL;
```
Why is obj->handler set to NULL just after being called ?
will it be set again by the I2C driver or application ?

LMESTM reviewed (commented):
Few comments or questions  ...

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -258,7 +265,77 @@ void i2c_sw_reset(i2c_t *obj)
+            DEBUG_PRINTF("TIMEOUT or error in i2c_slave_read\r\n");
+        }
+    }
+    return count;
```
Shouldn't be all return values based on 'handle->XferCount' to reflect real sent/received count

Commit:
docs(hal-i2c): Update design document to match recent API alterations

Commit:
fix(hal-i2c): STM: Fix frequency function checking the wrong variable

The frequency function checks the frequency requested by the caller,
rather than the actual supported frequency selected by the function.

Commit:
style(hal-i2c): STM: Remove empty comment block from code

Commit:
fix(hal-i2c): STM: Add missing return statement when re-initing as slave

- Stop function should exit early if the device is reinitialised in
  slave mode.

- Remove slave mode lookup when reinitialising, this block is only hit
  if the peripheral is in slave mode, we don't need to check again.

Commit:
refactor(hal-i2c): Add an API function to configure clock stretching

- Remove clock stretching behavior from `i2c_timeout` function and
  define a new function to enable/disable clock stretching behavior.

- Add implementation for STM32 boards and track configuration in the
  platform specific I2C struct.

Commit:
refactor(hal-i2c): Remove i2c_timeout function from the API

0xc0170 commented:
@scartmell-arm Is it time to update this one and make sure reviewers find time to review?

ghost commented:
> @scartmell-arm Is it time to update this one and make sure reviewers find time to review?

There were a few outstanding concerns I was working through, I'll push a resolution to all of them soon.

Commit:
style(hal-i2c): Fix astyle code formatting failures

Commit:
fix(hal-i2c): K64F: Ignore calls to set slave address in master mode

maciejbocianski commented on hal/i2c_api.h:
```diff
@@ -67,84 +82,162 @@ enum {
+ *                otherwise disable it.
  */
-int  i2c_stop(i2c_t *obj);
+void i2c_set_clock_stretching(i2c_t *obj, const bool enabled);
```
I think `const` is redundant here

0xc0170 commented:
@scartmell-arm Who should review , anyone from @ARMmbed/mbed-os-hal  or @maciejbocianski is sufficient (still some comments not addressed).

How can we progress i2c feature branch?

Commit:
fix(hal-i2c): Disable `I2CSLAVE` feature flag on K64F devices

donatieng commented on drivers/I2C.cpp:
```diff
@@ -191,40 +189,17 @@ int I2C::recover(PinName sda, PinName scl)
 
 int I2C::transfer(int address, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length, const event_callback_t &callback, int event, bool repeated)
 {
```
I think that shouldn't be empty!

donatieng reviewed (changes requested):
Missing implementation of async in the driver - @scartmell-arm @maciejbocianski would you be able to pick this up so that we get this merged soon?

Commit:
fix(hal-i2c): Return the num bytes transferred in STM slave read/write

Commit:
fix(hal-i2c): Disable slave mode and 10-bit from I2C capabilities on K64

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -508,10 +634,15 @@ int i2c_stop(i2c_t *obj)
+    }
 }
 
 int i2c_byte_read(i2c_t *obj, int last)
```
@scartmell-arm 
`i2c_byte_read` and  `i2c_byte_write` should be removed since it was removed from API

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -258,7 +265,78 @@ void i2c_sw_reset(i2c_t *obj)
+    capabilities->minimum_frequency = MINIMUM_FREQUENCY;
+    capabilities->maximum_frequency = MAXIMUM_FREQUENCY;
+    capabilities->supports_slave_mode = true;
+    capabilities->supports_10bit_addressing = true;
```
@scartmell-arm also here 10bit addressing should be disabled

Commit:
fix(hal-i2c): Disable 10-bit addressing I2C capabilities on STM32

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -777,37 +909,47 @@ int i2c_read(i2c_t *obj, int address, char *data, int length, int stop)
-    return count;
+    i2c_ev_err_disable(obj);
+
+    if ((timeout == 0) || (obj_s->event != I2C_EVENT_TRANSFER_COMPLETE)) {
```
@scartmell-arm 
I think that inside this `if` we should detect error type and return appropriately

maciejbocianski commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -831,36 +973,45 @@ int i2c_write(i2c_t *obj, int address, const char *data, int length, int stop)
+
+    i2c_ev_err_disable(obj);
+
+    if ((timeout == 0) || (obj_s->event != I2C_EVENT_TRANSFER_COMPLETE)) {
```
@scartmell-arm 
I think that inside this `if` we should detect error type and return appropriately

Commit:
fix(hal-i2c): Fix compiler warning from unused variable on STM32

Commit:
fix(hal-i2c): Fix return value for transfers that timeout in STM

maciejbocianski committed:
rebase fixes
